[WIP] Add support for `approx.auto_quantize`

bushshrub opened the pull request:
This PR adds support for `approx.auto_quantize` as covered in #16.

This function takes some model (currently only torch module) and returns a quantized version of that model

Also changes the function signature of `auto_set_backend()` to modify a library variable instead

Closes: #16 

May also close #3

bushshrub committed:
Remove context manager

This was not designed to be used in such a manner

bushshrub committed:
Add support for auto_set_backend and add auto_quantize API

bushshrub committed:
Update example to demonstrate usage with auto_quantization on a model

bushshrub committed:
Search for preferred backends first before falling back

bushshrub committed:
Update CI workflow to run tests

bushshrub committed:
Fix tests for auto_set_backend()

bushshrub committed:
Fix lint issues

bushshrub committed:
Update lint workflow to run checks

bushshrub committed:
Rename CI workflow

bushshrub committed:
Rename documentation workflow from CI workflow

bushshrub committed:
Fix wrong type annotations

bushshrub committed:
Update documentation to warn about removal of auto_cast_all

bushshrub committed:
check if _APPROX_BACKEND is set before operating

ma7dev committed:
feat(precommit): Adding mypy and publish ci

- Added mypy to the project and to pre-commit to ensure the linting passes
    before every commit.
- Run mypy on all files and fixed the linting problems.
- Included publish to PyPi using a ci

Resolve: #14 #15

xiurobert commented on examples/scripts/mnist_torch.py:
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
-loss = nn.CrossEntropyLoss().to("cuda")
+    model = BasicNN()
+    model = approx.auto_quantize(model, pretrained=True)
 
```
Nit: extra line spacing

bushshrub replied:
Extra spacing

bushshrub commented on examples/scripts/mnist_torch.py:
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
     for epoch_it in range(num_epochs):
         prog = tqdm(enumerate(train_loader), total=len(train_loader))
         for batch_num, (features, labels) in prog:
+            # approx.auto_cast_all(features, labels)
```
Remove unnecessary comment

bushshrub commented on examples/scripts/mnist_torch.py:
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
 
-with approx.auto_cast_all():
-    num_epochs = 10
+    # qmodel = torch.quantization.quantize_dynamic(model, {nn.Linear, nn.Conv2d}, dtype=torch.qint8)
```
Remove this

bushshrub commented on examples/scripts/mnist_torch.py:
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
-with approx.auto_cast_all():
-    num_epochs = 10
+    # qmodel = torch.quantization.quantize_dynamic(model, {nn.Linear, nn.Conv2d}, dtype=torch.qint8)
+    # qmodel.to('cuda')
+    print(model)
+    exit(0)
+
+    # model.to("cuda")
```
Fix this

bushshrub commented on tests/integration/core/test_api.py:
```diff
@@ -5,12 +5,14 @@
     """Tests that the auto_set_backend function works as expected."""
-    from approx.core.api import auto_set_backend
-    from approx.core.backend._backend import BackendType
+    # from approx.core.api import auto_set_backend
```
Unnecessary

bushshrub commented on approx/core/backend/_backend.py:
```diff
@@ -26,30 +28,60 @@ def __init__(self, backend):
+        """
+        if pretrained:
+            qmodel = self.module.quantization.quantize_dynamic(
+                model,
+                {
+                    # todo: this is definitely not enough
+                    self.module.nn.Linear,
+                    self.module.nn.Conv2d,
+                },
+                dtype=self.module.qint8,
```
More quantization options?

@sudomaze

ma7dev replied:
I am adding more per [1](https://pytorch.org/docs/stable/_modules/torch/ao/quantization/quantize.html#quantize_dynamic) and [2](https://pytorch.org/docs/stable/quantization.html#quantization-customizations). Here is the current list that was added: 

```python
    self.module.nn.Linear,
    self.module.nn.LSTM,
    self.module.nn.GRU,
    self.module.nn.LSTMCell,
    self.module.nn.RNNCell,
    self.module.nn.GRUCell,
    self.module.nn.Conv2d,
    self.module.nn.Conv3d,
```

I will add some unit tests to ensure the quantization is working as expected. We will need to do a similar mapping in other backends.

bushshrub replied:
SGTM. I was mainly working on PyTorchBackend until we get that decently well working, before we support others

bushshrub reviewed (commented):
Few more things

bushshrub committed:
Fix faulty example

bushshrub committed:
Resolve some comments

ma7dev committed:
[WIP] fix(backend): Move to CPU when on GPU

ma7dev committed:
feat(backend): Added quantized modules with tests

- Updated quantized modules in backend
- Updated `CONTRIBUTING.md` to enable `pre-commit` to be run
    automatically without running it manually.
- Added `test_backend` to test backend modules

TODO:
- [ ] Test while using `Sequential`
- [ ] Test while having more nested/complex modules
- [ ] Test when giving one operation only

ma7dev committed:
fix(pre-commit): Run pre-commit on all of files

bushshrub commented:
Blocking, waiting for #22 to hit develop

bushshrub committed:
Drop `auto_cast_all`

bushshrub committed:
Forgot to remove it from import

bushshrub committed:
Drop `auto_cast_all`

bushshrub committed:
Merge branch 'develop' into dev/test_backend_pytorch

ma7dev reviewed (approved):
<img src="https://media3.giphy.com/media/3o6UB0WHErPubNDTfa/giphy.gif"/>
